Particle Overview (#291)
MinecraftForge/Documentation: docs/effects/particles.md

Particles
=========

Terminology
-----------
| Term | Description |
|----------------|----------------|
| Particle | What is responsible for the rendering logic. |
| ParticleType<?> | What is registered using a unique name. |
| IParticleData | What contains server data that needs to be transferred to the client. |
| IParticleFactory | The factory takes in an IParticleData and returns a Particle. <br> It will be registered using a unique ParticleType. |

Different Particles

> -------------------

Reviewer: Your section headers should all have a blank line underneath.

> There are 4 abstract particle classes: Particle, TexturedParticle,  SpriteTexturedParticle and SimpleAnimatedParticle. Each inherit from the previous and your own Particle has to extend one of these.

Reviewer: Most of this line goes without saying, and can be discovered with IDE tools.

> In it, the rendering of the Particle is handled. I suggest looking at vanilla examples on how to handle rendering, this guide will not get into that, only the logic part of Particles.

Reviewer: I generally prefer the collective first person "We" as this is a collaborative project speaking from the perspective of a team. But it's not super important.

### Sidedness

The Particle class is [**client only**][sides], it does not exist on the server.
This means that creating a fixed parameter particle (no context information needed) is somewhat simpler, more on that later. 

> However, if the particle has a parameter with continuous values, server information is required. For example, breaking blocks create a particle effect based on the broken block's texture.

Reviewer: "continuous" is not the right word here. "dynamic" perhaps?

> This information obviously depends on the BlockState of the block, which is taken from the server. <br>

Reviewer: Avoid html tags. Double line break will accomplish the same thing.

> When adding a particle on the client, use ClientWorld#addParticle, but on the server it is ServerWorld#spawnParticle. ServerWorld#addParticle will simply do nothing (careful, one overload of the method in World is ClientOnly). 

Reviewer: ServerWorld#addParticle will simply do nothing ... when? Not always, surely. Also, try to put class/method names in backticks.

Author: World#addParticle does nothing, only ClientWorld overrides it to do something. Which is why i mention spawnParticle, which sends a particle spawn packet.

### Does your particle need server data?
Say you want to emit a particle on spell use. If you have 8 type of spells with each one specific color no need for server information. Each spell will have a distinct registered ParticleType. 

> However, if these spells can have modifiers that can affect the color in multiple ways (add/subtract red/green/blue data) then having a registry object for each combination will be fastidious. 

Reviewer: I'm not sure "fastidious" makes sense here.

Author: yeah my bad, its similar to a word in French in pronunciation but not in meaning, changed to tedious.

In that case, server information will be required.

> Meaning when spawning the particle you calculate the color to pass in instead of spawning a specific particle based on which spell was used.

Reviewer: These two sentences can be one.

| `IParticleFactory` | The factory takes in an `IParticleData` and returns a Particle. It will be registered using a unique `ParticleType<?>`. |

Different Particles
-------------------

You will need to have your own implementation of the `Particle` class, which will handle the rendering of the particle. Vanilla examples will help you on how to handle rendering, this guide will not get into that and only discuss the logic part of Particles.

### Sidedness
The `Particle` class is [**client only**][sides], it does not exist on the server. This means that creating a fixed parameter particle (no context information needed) is somewhat simpler, more on that later. However, if the particle has a parameter with dynamic values, server information is required. For example, breaking blocks create a particle effect based on the broken block's texture. This information obviously depends on the BlockState of the block, which is taken from the server.

To make a particle appear, use either `ClientWorld#addParticle` or
`ServerWorld#spawnParticle`. `ServerWorld#addParticle` will simply do nothing, without throwing any errors.

### Does your particle need server data?

> Taking as an example a mod which adds some spells to the game. f there was 8 type of spells added with each one specific color, no need for server information. Each spell will have a distinct registered `ParticleType`. 

Reviewer: ```suggestion
Taking as an example a mod which adds some spells to the game: If there's 8 type of spells added with each one specific color, there is no need for server information. Each spell will have a distinct registered `ParticleType`. 
```

However, if these spells can have modifiers that can affect the color in multiple ways (add/subtract red/green/blue data) then having a registry object for each combination will be tedious. In that case, server information will be required when spawning the particle, to calculate the color to use.

ParticleTypes
-------------

#### Vanilla Implementation
While there are a lot of different Particles in vanilla, in almost all cases vanilla uses BasicParticleType, a basic implementation of ParticleType and IParticleData. This implementation is used for anything that does not require [server data][servdat]. The only vanilla particles that do not use BasicParticleType are redstone dust and block/item texture dependent particles. When requiring server data, a direct implementation of IParticleData is needed. 

The `ParticleTypes` class is very helpful to check out vanilla implementations. Seeing how vanilla implemented each particle will help better evaluate what you need.

#### Registering

> ParticleType<?> is a ForgeRegistryEntry, so it is [registered normally][registration]. A `ParticleType` needs to be registered for each distinct `Particle`. When using server information, only one `ParticleType` needs to be registered, since it will cover all cases. When opting for covering all discrete cases, multiple ParticleType are registered all using `BasicParticleType`. The difference will come in registering the `IParticleFactory`.

Reviewer: Still some missing backticks
